### PR TITLE
Add azure-cli-managedservices module

### PR DESCRIPTION
**Pull request by rajivnandivada**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

File: azure-cli2017.pyproj [hunk 1]
```diff
@@ -13,8 +13,7 @@
     <LaunchProvider>Standard Python launcher</LaunchProvider>
     <InterpreterId>MSBuild|env|$(MSBuildProjectFullPath)</InterpreterId>
     <EnableNativeCodeDebugging>False</EnableNativeCodeDebugging>
-    <CommandLineArguments>
-    </CommandLineArguments>
+    <CommandLineArguments>managedservices definition assignment show --assignment "27fb250b-fdd1-4f2c-87da-7967e6028200"</CommandLineArguments>
```
Comment (tjprescott) [inline azure-cli2017.pyproj]: Revert.

File: doc/sphinx/azhelpgen/doc_source_map.json [hunk 2]
```diff
@@ -67,5 +67,6 @@
     "aks": "src/command_modules/azure-cli-acs/azure/cli/command_modules/acs/_help.py",
     "deployment": "src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_help.py",
     "functionapp": "src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_help.py",
-    "lock": "src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_help.py"
+    "lock": "src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_help.py",
+	"managedservices" : "src/command_modules/azure-cli-managedservices/azure/cli/command_modules/managedservices/_help.py"
```
Comment (tjprescott) [inline doc/sphinx/azhelpgen/doc_source_map.json]: Fix indent.

File: src/command_modules/azure-cli-managedservices/azure/cli/command_modules/managedservices/__init__.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+from ._command_type import ManagedServicesCommandGroup
+
+
+class ManagedServicesCommandsLoader(AzCommandsLoader):
+
+    def __init__(self, cli_ctx=None):
+        from azure.cli.core.commands import CliCommandType
+        managedservices_custom = CliCommandType(operations_tmpl='azure.cli.command_modules.managedservices.custom#{}')
+        super(ManagedServicesCommandsLoader, self).__init__(cli_ctx=cli_ctx,
+                                                            min_profile='2017-03-10-profile',
```
Comment (tjprescott) [inline src/command_modules/azure-cli-managedservices/azure/cli/command_modules/managedservices/__init__.py]: @limingu please provide guidance on the right way to annotate this. `min_profile` has been removed.

File: src/command_modules/azure-cli-managedservices/azure/cli/command_modules/managedservices/_command_type.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+    def _register_common_arguments(self, name, method_name=None, **kwargs):
+        command = self.command_loader.command_table[self.custom_command(name, method_name, **kwargs)]
+
+        group_name = 'ManagedServices'
+        command.add_argument('api_version', required=False, default=None,
+                             arg_group=group_name, help='The API Version to target.')
+        command.add_argument('subscription', '--subscription', '-s', required=False, default=None,
+                             arg_group=group_name,
+                             help='Optional, but can be used to override the default subscription.')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-managedservices/azure/cli/command_modules/managedservices/_command_type.py]: This already exists. Clobbering this risks breaking the functionality users expect when using `--subscription` (for example: using the subscription name or GUID).

File: src/command_modules/azure-cli-managedservices/azure/cli/command_modules/managedservices/_command_type.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+        if "assignment" in method_name:
+            command.add_argument('resource_group_name', '--resource-group', '-g', arg_group=group_name,
+                                 help='Optional. When provided the assignment will be created under the resource '
+                                      'group scope. Ex: /subscriptions/id/resourceGroups/rgName/.')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-managedservices/azure/cli/command_modules/managedservices/_command_type.py]: Again, `resource_group_name` is globally aliased. Additionally, you should not put "Required" or "Optional" in help text. The CLI already prints a status tag for required properties.

File: src/command_modules/azure-cli-managedservices/azure/cli/command_modules/managedservices/_help.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+"""
+
+helps['managedservices definition create'] = """
+    type: command
+    short-summary: Creates a new registration definition.
+    long-summary: Creates a new registration definition under the default scope or under the subscription provided.
+    parameters:
+        - name: --name -n
+          long-summary: |
+             The name of the registration definition.
```
Comment (tjprescott) [inline src/command_modules/azure-cli-managedservices/azure/cli/command_modules/managedservices/_help.py]: In all likelihood these should all be short summaries.

**Comment by rajivnandivada** [comment]
> Needs to go through command review.

@tjprescott - @limingu helped me the command review. This PR includes that feedback. Thank you.

File: azure-cli2017.pyproj (hunk 1 above)
Comment (rajivnandivada) [reply azure-cli2017.pyproj]: Sorry about that. Reverted

File: src/command_modules/azure-cli-managedservices/azure/cli/command_modules/managedservices/__init__.py (hunk 3 above)
Comment (limingu) [reply src/command_modules/azure-cli-managedservices/azure/cli/command_modules/managedservices/__init__.py]: Remove the line of min_profile='2017-03-10-profile',

Comment (limingu) [reply src/command_modules/azure-cli-managedservices/azure/cli/command_modules/managedservices/__init__.py]: Create ResourceType for ManagedServices such as MGMT_MANAGEDSERVICES and add as resource_type=ResourceType.MGMT_MANAGEDSERVICES ,
https://github.com/Azure/azure-cli/blob/dev/src/azure-cli-core/azure/cli/core/profiles/_shared.py#L33

**Comment by marstr** [comment]
Howdy @rajivnandivada, it looks like you got caught at a particularly bad time by #9714. My GitHub handle matches my internal alias. Please reach out for help integrating your changes.

**Comment by marstr** [comment]
Looks like the history file is conflicting with changes that have gone into [`dev`](https://github.com/Azure/azure-cli/tree/dev) since your first merge commit! Should be pretty trivial to fix it up, so I'll let you handle that :)

**Comment by rajivnandivada** [comment]
@yugangw-msft would you please review this? Thank you.

File: azure-cli2017.pyproj [hunk 7]
```diff
@@ -14,7 +14,7 @@
     <InterpreterId>MSBuild|env|$(MSBuildProjectFullPath)</InterpreterId>
     <EnableNativeCodeDebugging>False</EnableNativeCodeDebugging>
     <CommandLineArguments>
-    </CommandLineArguments>
+	</CommandLineArguments>
```
Comment (limingu) [inline azure-cli2017.pyproj]: Keep the old version. You do need to change that.

File: azure-cli2017.pyproj [hunk 8]
```diff
@@ -1769,16 +1786,7 @@
...
-      <Id>env2</Id>
-      <Version>2.7</Version>
-      <Description>env2 (Python 2.7 (64-bit))</Description>
-      <InterpreterPath>Scripts\python.exe</InterpreterPath>
-      <WindowsInterpreterPath>Scripts\pythonw.exe</WindowsInterpreterPath>
-      <PathEnvironmentVariable>PYTHONPATH</PathEnvironmentVariable>
-      <Architecture>X64</Architecture>
-    </Interpreter>
-    <Interpreter Include="..\..\env\">
+    <Interpreter Include="..\env\">
```
Comment (limingu) [inline azure-cli2017.pyproj]: Keep the old version. You do need to change that.

File: src/azure-cli-core/azure/cli/core/profiles/_shared.py [hunk 9]
```diff
@@ -199,7 +200,8 @@ def default_api_version(self):
             'classic_administrators': '2015-06-01'
         }),
         ResourceType.DATA_KEYVAULT: '2016-10-01',
-        ResourceType.DATA_STORAGE: '2015-04-05'
+        ResourceType.DATA_STORAGE: '2015-04-05',
+        ResourceType.MGMT_MANAGEDSERVICES: '2018-06-01-preview'
```
Comment (limingu) [inline src/azure-cli-core/azure/cli/core/profiles/_shared.py]: You only need to add MGMT_MANAGEDSERVICES here only if 2017-03-09-profile needs to support managed service commands.

File: src/azure-cli/HISTORY.rst [hunk 10]
```diff
@@ -34,6 +34,9 @@ Release History
 
 * Support replication for MariaDB.
 
+**Managed Services**
```
Comment (limingu) [inline src/azure-cli/HISTORY.rst]: Keep the alphabetic order.

File: src/azure-cli/setup.py [hunk 11]
```diff
@@ -95,6 +95,7 @@
     'azure-mgmt-kusto~=0.3.0',
     'azure-mgmt-loganalytics~=0.2',
     'azure-mgmt-managementgroups~=0.1',
+    'azure-mgmt-managedservices~=0.1',
```
Comment (limingu) [inline src/azure-cli/setup.py]: Keep the alphabetic order

File: src/azure-cli-core/azure/cli/core/profiles/_shared.py (hunk 9 above)
Comment (rajivnandivada) [reply src/azure-cli-core/azure/cli/core/profiles/_shared.py]: I see. I mistook it for the api version.

**Comment by rajivnandivada** [comment]
> Needs to go through command review.

Done.

**Comment by rajivnandivada** [comment]
@limingu I have addressed your feedback. Please review. Thank you.

File: azure-cli2017.pyproj [hunk 12]
```diff
@@ -1220,6 +1232,11 @@
     <Folder Include="command_modules\azure-cli-ams\azure\cli\command_modules\" />
     <Folder Include="command_modules\azure-cli-ams\azure\cli\command_modules\ams\" />
     <Folder Include="command_modules\azure-cli-ams\azure\cli\command_modules\ams\tests\" />
+    <Folder Include="command_modules\azure-cli-managedservices\" />
+    <Folder Include="command_modules\azure-cli-managedservices\azure\" />
+    <Folder Include="command_modules\azure-cli-managedservices\azure\cli\" />
+    <Folder Include="command_modules\azure-cli-managedservices\azure\cli\command_modules\" />
+    <Folder Include="command_modules\azure-cli-managedservices\azure\cli\command_modules\managedservices\" />
```
Comment (tjprescott) [inline azure-cli2017.pyproj]: These includes will be broken with the new package structure. Please update.

File: src/azure-cli/azure/cli/command_modules/managedservices/_command_type.py [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+        command = self.command_loader.command_table[self.custom_command(name, method_name, **kwargs)]
+
+        group_name = 'ManagedServices'
+        command.add_argument('api_version', required=False, default=None,
+                             arg_group=group_name, help='The API Version to target.')
+
+        if method_name in ("cli_assignment_list", "cli_assignment_get"):
+            from azure.cli.core.commands.parameters import get_three_state_flag
+            command.add_argument('include_definition', arg_group=group_name, arg_type=get_three_state_flag(),
+                                 help='When provided, gets the associated registration definition details.')
```
Comment (tjprescott) [inline src/azure-cli/azure/cli/command_modules/managedservices/_command_type.py]: The standard way to register this would be for `cli_assignment_list` and `cli_assignment_get` to both have `include_definition` in their signature and then have a simple `c.argument('include_definition', …)` line in params.py. Is there a reason you are doing it this way?

File: src/azure-cli/azure/cli/command_modules/managedservices/_help.py [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+             The managed by tenant id.
+        - name: --principal-id
+          short-summary: |
+             The principal id.
+        - name: --role-definition-id
+          short-summary: |
+             The role definition id.
+        - name: --description
+          short-summary: |
+             Optional: The friendly description.
```
Comment (tjprescott) [inline src/azure-cli/azure/cli/command_modules/managedservices/_help.py]: Remove any references to "Required:" or "Optional:". The absence of the CLI's automatic `[Required]` tag implies that something is optional.

File: src/azure-cli/azure/cli/command_modules/managedservices/_help.py [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+"""
+
+helps['managedservices definition delete'] = """
+    type: command
+    short-summary: Deletes a registration definition.
+    long-summary: Deletes a registration definition when given its identifier or the fully qualified resource id. When resource id is provided, the subscription paramter is ignored.
+    parameters:
+        - name: --definition
+          short-summary: |
+             The identifier (guid) or the fully qualified resource id of the registration definition. When resource id is used, subscription id and resource group parameters are ignored.
```
Comment (tjprescott) [inline src/azure-cli/azure/cli/command_modules/managedservices/_help.py]: There is a lot of duplication here that could make maintenance difficult or at least irritating. Typically, these short summaries would simply be registered at the appropriate scope in params.py to avoid this duplication. Use of help.py for parameters is only necessary in the event long summaries or populator commands are needed (features not available in params.py).

File: src/azure-cli/azure/cli/command_modules/managedservices/_help.py [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+"""
+
+helps['managedservices definition assignment create'] = """
+    type: command
+    short-summary: Creates a new registration assignment.
+    long-summary: Creates a new registration assignment when given the fully qualified resource id of the registration definition.
+    parameters:
+        - name: --definition-id
+          short-summary: |
+             The fully qualified resource id of the registration definition.
```
Comment (tjprescott) [inline src/azure-cli/azure/cli/command_modules/managedservices/_help.py]: If this is an arm resource, it should be `--definition` and should follow the logic that `--definition` does in other commands.

File: src/azure-cli/azure/cli/command_modules/managedservices/_help.py [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+        - name: Deletes an assignment given its identifier under the default subscription scope.
+          text: az managedservices definition assignment delete --assignment d3087cf0-e180-4cca-b147-54ae00c7b504
+        - name: Deletes an assignment given its fully qualified resource id.
+          text: az managedservices definition assignment delete --assignment /subscriptions/a62076fa-768a-403c-9d9d-6a9919aae441/providers/Microsoft.ManagedServices/registrationAssignments/0c3e9687-b461-4615-b6e4-74d54998d6e4
+"""
+
+helps['managedservices definition assignment list'] = """
+    type: command
+    short-summary: List all the registration assignments.
+    long-summary: List all the registration assignments. Subscription id and resource group parameters can be used to override default values.
```
Comment (tjprescott) [inline src/azure-cli/azure/cli/command_modules/managedservices/_help.py]: The long summary should not duplicate the short summary. The CLI essentially concatenates the two so this will result in duplicate lines.

File: src/azure-cli/azure/cli/command_modules/managedservices/commands.py [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+        operations_tmpl='azure.mgmt.managedservices.operations#RegistrationAssignmentsOperations.{}',
+        client_factory=cf_registration_assignments)
+
+    with self.command_group('managedservices definition', msp_registration_definitions) as g:
+        g.managedservices_custom('create', 'cli_definition_create', client_factory=cf_registration_definitions)
+        g.managedservices_custom('list', 'cli_definition_list', client_factory=cf_registration_definitions)
+        g.managedservices_custom('delete', 'cli_definition_delete', client_factory=cf_registration_definitions)
+        g.managedservices_custom('show', 'cli_definition_get', client_factory=cf_registration_definitions)
+
+    with self.command_group('managedservices definition assignment', msp_registration_assignments) as g:
```
Comment (tjprescott) [inline src/azure-cli/azure/cli/command_modules/managedservices/commands.py]: Highly recommend this follow the same convention as `az role` commands:
```
Group
    az managedservices : Help text...

Subgroups:
    assignment : Manage definition assignments.
    definition : Manage definitions.
```

Beyond consistency, this flatter structure results in shorter commands and makes these the assignment commands more discoverable.

File: src/azure-cli/azure/cli/command_modules/managedservices/commands.py [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+        g.managedservices_custom('create', 'cli_assignment_create', client_factory=cf_registration_assignments)
+        g.managedservices_custom('show', 'cli_assignment_get', client_factory=cf_registration_assignments)
+        g.managedservices_custom('delete', 'cli_assignment_delete', client_factory=cf_registration_assignments)
+        g.managedservices_custom('list', 'cli_assignment_list', client_factory=cf_registration_assignments)
```
Comment (tjprescott) [inline src/azure-cli/azure/cli/command_modules/managedservices/commands.py]: Since these commands are in preview, you need to add the following:

```Python
with self.command_group('managedservices', is_preview=True) as g:
   pass
```

This will ensure that all of your commands are appropriately marked as preview.

File: src/azure-cli/azure/cli/command_modules/managedservices/custom.py [hunk 20]
```diff
@@ -0,0 +1,219 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+import uuid
+from knack.log import get_logger
+from msrestazure.tools import parse_resource_id, is_valid_resource_id
+
+logger = get_logger(__name__)
+default_api_version = '2018-06-01-preview'
```
Comment (tjprescott) [inline src/azure-cli/azure/cli/command_modules/managedservices/custom.py]: Is the API version for Managed Services, or an input for the API version of the interfacing resources?

File: src/azure-cli/azure/cli/command_modules/managedservices/custom.py [hunk 21]
```diff
@@ -0,0 +1,219 @@
...
+    scope = _get_scope(sub_id)
+    return client.list(
+        scope=scope,
+        api_version=_get_api_version(api_version))
+
+
+def cli_definition_delete(cmd, client,
+                          definition, **kwargs):
+    subscription = kwargs.pop('subscription', None)
+    api_version = kwargs.pop('api_version', None)
```
Comment (tjprescott) [inline src/azure-cli/azure/cli/command_modules/managedservices/custom.py]: Why are you using kwargs?  These arguments should just be in the signature for the command.

File: src/azure-cli/azure/cli/command_modules/managedservices/custom.py [hunk 22]
```diff
@@ -0,0 +1,219 @@
...
+
+# pylint: disable=unused-argument
+def cli_assignment_create(cmd, client,
+                          definition_id,
+                          assignment_id=None,
+                          **kwargs):
+    from azure.mgmt.managedservices.models import RegistrationAssignmentProperties
+    if not is_valid_resource_id(definition_id):
+        raise ValueError(
+            "registration_definition_resource_id should be a valid resource id. For example, "
```
Comment (tjprescott) [inline src/azure-cli/azure/cli/command_modules/managedservices/custom.py]: "registration_definition_resource_id" should correspond to the actual option, i.e. `--definition-id` or whatever it is.

File: src/azure-cli/azure/cli/command_modules/managedservices/tests/latest/test_managedservices_scenario.py [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+        result = self.cmd(
+            'az managedservices definition create --name {name} --tenant-id  {tenant-id} --principal-id {principal-id} --role-definition-id {role-definition-id} --definition-id {definition-id} --subscription {subscription-id}').get_output_in_json()
+
+        self.assertTrue(result['name'], definition_id)
+        self.assertTrue(result['properties'] is not None)
+        self.assertTrue(result['properties']['provisioningState'], "Succeeded")
+        self.assertTrue(result['properties']['managedByTenantId'], tenant_id)
+        self.assertTrue(result['properties']['authorizations'] is not None)
+        self.assertTrue(result['properties']['authorizations'][0]['roleDefinitionId'], role_definition_id)
+        self.assertTrue(result['properties']['authorizations'][0]['principalId'], principal_id)
```
Comment (tjprescott) [inline src/azure-cli/azure/cli/command_modules/managedservices/tests/latest/test_managedservices_scenario.py]: The recommended way of doing these would be:

```Python
self.cmd('az managedservices definition create --name {name} --tenant-id  {tenant-id} --principal-id {principal-id} --role-definition-id {role-definition-id} --definition-id {definition-id} --subscription {subscription-id}', checks=[
  self.check('name', '{definition-id}'),
  self.check('properties.provisioningState', 'Succeeded'),
  ...
]) 
```

More concise and generally results in better error messages when a test fails (I believe these will yield the ultra helpful "False is not True").

File: azure-cli2017.pyproj (hunk 12 above)
Comment (rajivnandivada) [reply azure-cli2017.pyproj]: Sure. The cli.pyproj needs to be updated after the structure change but I have moved managedservices to the appropriate location in the solution.

File: src/azure-cli/azure/cli/command_modules/managedservices/custom.py (hunk 21 above)
Comment (rajivnandivada) [reply src/azure-cli/azure/cli/command_modules/managedservices/custom.py]: I was following the example in the Storage but I will change.

File: src/azure-cli/setup.py [hunk 24]
```diff
@@ -53,97 +53,66 @@
 ]
 
 DEPENDENCIES = [
```
Comment (marstr) [inline src/azure-cli/setup.py]: Please undo all changes to our dependency list.

File: doc/sphinx/azhelpgen/doc_source_map.json [hunk 25]
```diff
@@ -1,72 +1,72 @@
...
-    "eventhubs": "src/azure-cli/azure/cli/command_modules/eventhubs/_help.py",
-    "servicebus": "src/azure-cli/azure/cli/command_modules/servicebus/_help.py",
-    "search": "src/azure-cli/azure/cli/command_modules/search/_help.py",
-    "managedapp": "src/azure-cli/azure/cli/command_modules/resource/_help.py",
-    "aks": "src/azure-cli/azure/cli/command_modules/acs/_help.py",
-    "deployment": "src/azure-cli/azure/cli/command_modules/resource/_help.py",
-    "functionapp": "src/azure-cli/azure/cli/command_modules/appservice/_help.py",
-    "lock": "src/azure-cli/azure/cli/command_modules/resource/_help.py",
-    "managedservices": "src/azure-cli/azure/cli/command_modules/managedservices/_help.py"
+    "az": "src/command_modules/azure-cli-profile/azure/cli/command_modules/profile/_help.py",
```
Comment (marstr) [inline doc/sphinx/azhelpgen/doc_source_map.json]: The reason there are so many edits to this file is because of a merge error. Please undo all changes that are not related to your module.

File: src/azure-cli/azure/cli/command_modules/managedservices/_command_type.py (hunk 13 above)
Comment (rajivnandivada) [reply src/azure-cli/azure/cli/command_modules/managedservices/_command_type.py]: Fixed.

File: src/azure-cli/azure/cli/command_modules/managedservices/_help.py (hunk 17 above)
Comment (rajivnandivada) [reply src/azure-cli/azure/cli/command_modules/managedservices/_help.py]: Understood. I changed all the long summaries to short and remove the long ones.

File: src/azure-cli/azure/cli/command_modules/managedservices/commands.py (hunk 18 above)
Comment (rajivnandivada) [reply src/azure-cli/azure/cli/command_modules/managedservices/commands.py]: Done.

File: src/azure-cli/azure/cli/command_modules/managedservices/commands.py (hunk 19 above)
Comment (rajivnandivada) [reply src/azure-cli/azure/cli/command_modules/managedservices/commands.py]: Done.

File: src/azure-cli/azure/cli/command_modules/managedservices/custom.py (hunk 20 above)
Comment (rajivnandivada) [reply src/azure-cli/azure/cli/command_modules/managedservices/custom.py]: This is the API version for managed services.

File: src/azure-cli/azure/cli/command_modules/managedservices/custom.py (hunk 21 above)
Comment (rajivnandivada) [reply src/azure-cli/azure/cli/command_modules/managedservices/custom.py]: Remove kwargs.

File: src/azure-cli/azure/cli/command_modules/managedservices/custom.py (hunk 22 above)
Comment (rajivnandivada) [reply src/azure-cli/azure/cli/command_modules/managedservices/custom.py]: changed to definition

File: src/azure-cli/azure/cli/command_modules/managedservices/tests/latest/test_managedservices_scenario.py (hunk 23 above)
Comment (rajivnandivada) [reply src/azure-cli/azure/cli/command_modules/managedservices/tests/latest/test_managedservices_scenario.py]: done.

File: azure-cli2017.pyproj [hunk 26]
```diff
@@ -598,6 +608,18 @@
...
+    <!-- <Compile Include="command_modules\azure-cli-managedservices\azure\cli\command_modules\managedservices\tests\latest\test_managedservices_scenario.py" /> -->
+    <Compile Include="command_modules\azure-cli-managedservices\azure\cli\command_modules\managedservices\_client_factory.py" />
+    <Compile Include="command_modules\azure-cli-managedservices\azure\cli\command_modules\managedservices\_command_type.py" />
+    <Compile Include="command_modules\azure-cli-managedservices\azure\cli\command_modules\managedservices\_help.py">
+      <SubType>Code</SubType>
+    </Compile>
+    <Compile Include="command_modules\azure-cli-managedservices\azure\cli\command_modules\managedservices\_params.py" />
+    <Compile Include="command_modules\azure-cli-managedservices\azure\cli\command_modules\managedservices\_validators.py" />
+    <Compile Include="command_modules\azure-cli-managedservices\azure\cli\command_modules\managedservices\__init__.py" />
+    <Compile Include="command_modules\azure-cli-managedservices\setup.py" />
```
Comment (tjprescott) [inline azure-cli2017.pyproj]: These entries need to be removed as these paths don't exist.

File: src/azure-cli/azure/cli/command_modules/managedservices/__init__.py [hunk 27]
```diff
@@ -0,0 +1,40 @@
...
+from azure.cli.command_modules.managedservices._help import helps  # pylint: disable=unused-import
+from azure.cli.core.commands import AzCommandGroup
+
+
+class ManagedServicesCommandGroup(AzCommandGroup):
+
+    def managedservices_custom(self, name, method_name=None, **kwargs):
+        command = self.command_loader.command_table[self.custom_command(name, method_name, **kwargs)]
+        if not command:
+            return
```
Comment (tjprescott) [inline src/azure-cli/azure/cli/command_modules/managedservices/__init__.py]: This custom command group does nothing and should be removed.

File: src/azure-cli/azure/cli/command_modules/managedservices/commands.py [hunk 28]
```diff
@@ -0,0 +1,34 @@
...
+
+    msp_registration_assignments = CliCommandType(
+        operations_tmpl='azure.mgmt.managedservices.operations#RegistrationAssignmentsOperations.{}',
+        client_factory=cf_registration_assignments)
+
+    with self.command_group('managedservices', is_preview=True) as g:
+        pass
+
+    with self.command_group('managedservices definition', msp_registration_definitions) as g:
+        g.managedservices_custom('create', 'cli_definition_create', client_factory=cf_registration_definitions)
```
Comment (tjprescott) [inline src/azure-cli/azure/cli/command_modules/managedservices/commands.py]: These should all just be `g.custom` since `managedservices_custom` doesn't add any functionality.

File: src/azure-cli/azure/cli/command_modules/managedservices/commands.py [hunk 29]
```diff
@@ -0,0 +1,34 @@
...
+        client_factory=cf_registration_definitions)
+
+    msp_registration_assignments = CliCommandType(
+        operations_tmpl='azure.mgmt.managedservices.operations#RegistrationAssignmentsOperations.{}',
+        client_factory=cf_registration_assignments)
+
+    with self.command_group('managedservices', is_preview=True) as g:
+        pass
+
+    with self.command_group('managedservices definition', msp_registration_definitions) as g:
```
Comment (tjprescott) [inline src/azure-cli/azure/cli/command_modules/managedservices/commands.py]: You can add `client_factory=cf_registration_definitions` here as a kwarg so that you don't have to add it to each of the subsequent command registrations.

File: src/azure-cli/azure/cli/command_modules/managedservices/custom.py (hunk 20 above)
Comment (tjprescott) [reply src/azure-cli/azure/cli/command_modules/managedservices/custom.py]: In that case, please remove. If you want your module to work with multiple APIs, there are several things you would need to do differently, and this is not the mechanism the CLI allows for that.

Comment (rajivnandivada) [reply src/azure-cli/azure/cli/command_modules/managedservices/custom.py]: @tjprescott could you point to an example where api version is handled correctly? thank you.

File: src/azure-cli/azure/cli/command_modules/managedservices/__init__.py (hunk 27 above)
Comment (rajivnandivada) [reply src/azure-cli/azure/cli/command_modules/managedservices/__init__.py]: removed

File: src/azure-cli/azure/cli/command_modules/managedservices/custom.py (hunk 20 above)
Comment (rajivnandivada) [reply src/azure-cli/azure/cli/command_modules/managedservices/custom.py]: I chatted with Travis offline regarding the API version. For now, I have created an issue to use a hardcoded value in the commands. I am going to follow up Laurent regarding the Python SDK.

File: src/azure-cli/azure/cli/command_modules/managedservices/custom.py [hunk 30]
```diff
@@ -7,7 +7,11 @@
 from msrestazure.tools import parse_resource_id, is_valid_resource_id
 
 logger = get_logger(__name__)
-default_api_version = '2018-06-01-preview'
+
+# The API version is not being set correctly in the generated Python SDK.
+# Opened an issue #https://github.com/Azure/azure-sdk-for-python/issues/6112 to track it.
+# Ideally, the SDK should set the version accordingly based on Swagger spec.
+API_VERSION_PREVIEW = '2018-06-01-preview'
```
Comment (tjprescott) [inline src/azure-cli/azure/cli/command_modules/managedservices/custom.py]: Also, add a reference here to the CLI issue:
```Python
# TODO: Remove when azure-cli issue #1234 is resolved.
```

**Comment by tjprescott** [comment]
LGTM. @limingu @marstr please re-review.

File: azure-cli2017.pyproj [hunk 31]
```diff
@@ -14,7 +14,7 @@
     <InterpreterId>MSBuild|env|$(MSBuildProjectFullPath)</InterpreterId>
     <EnableNativeCodeDebugging>False</EnableNativeCodeDebugging>
     <CommandLineArguments>
-    </CommandLineArguments>
+	  </CommandLineArguments>
```
Comment (limingu) [inline azure-cli2017.pyproj]: Wrong indent.

File: azure-cli2017.pyproj [hunk 32]
```diff
@@ -1769,16 +1784,7 @@
...
-      <Id>env2</Id>
-      <Version>2.7</Version>
-      <Description>env2 (Python 2.7 (64-bit))</Description>
-      <InterpreterPath>Scripts\python.exe</InterpreterPath>
-      <WindowsInterpreterPath>Scripts\pythonw.exe</WindowsInterpreterPath>
-      <PathEnvironmentVariable>PYTHONPATH</PathEnvironmentVariable>
-      <Architecture>X64</Architecture>
-    </Interpreter>
-    <Interpreter Include="..\..\env\">
+    <Interpreter Include="..\env\">
```
Comment (limingu) [inline azure-cli2017.pyproj]: You don't need to change that. Each developer may have different env setting.

Comment (rajivnandivada) [reply azure-cli2017.pyproj]: fixed.
